### ISSUE TITLE
Figure out how to version language plugins

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#956]
At the moment, although languages are semi-implemented as plugins, we don't yet have a versioning scheme for them.  And so we continue to bundle them in the SDK.  Eventually we want to support true independent versioning of them, especially to foster 3rd party extensibility here.

The challenge here is that the only metadata in `Pulumi.yaml` simply tells us whether the language is `nodejs`, `python`, etc.  I suspect we may want to allow a semantic version to go in there, e.g. `nodejs: ^0.10.0`, and to treat our current simple names as syntactic sugar for `latest`.

**Comment by swgillespie** [comment]
Do we still have a pressing need for this?

**Comment by lukehoban** [comment]
I'm not aware of anything that we've seen concretely require this yet.  Versioning resource providers (https://github.com/pulumi/pulumi/issues/2389) is much more critical currently as that is actively blocking our ability to safely deliver breaking changes across major version in providers (and in particular, to adopt new versions of TF providers which include breaking changes in major versions).

**Comment by swgillespie** [comment]
We still deliver language plugins alongside the CLI and so there is no need to change anything that we're doing vis-a-vis versioning at this time.

**Comment by ellismg** [comment]
I do not think this is a pressing need to do this in Q3.  Note that much of the engine already treats language plugins as special, since they are not treated the same was as resource provider plugins "(e.g. they don't end up in `~/.pulumi/plugins`).  As we evolve our multi-language strategy, I do believe we will need a way to do something like this, specifically when we want to stop being the only folks that ship language plugins, but for now, I don't think there's any work we need to track for Q3.

**Comment by lunaris** [comment]
Given how long we have not seen a pressing need to do this (at least, given my experience and outlook in the coming months), I'm going to close this as unplanned for now. We can reopen/pitch it as part of a future upcoming epic if need be.
